shef_loader/cda_loader.py

```diff
@@ -0,0 +1,393 @@
+import asyncio
+from datetime import datetime, timezone
+from io import BufferedRandom
+from itertools import groupby
+import json
+from logging import Logger
+import os
+import re
+import time
+from typing import (
+    Coroutine,
+    NamedTuple,
+    Optional,
+    TextIO,
+    Tuple,
+    TypedDict,
+    Union,
+    cast,
+)
+from shef_loader import shared
+from . import base_loader
+from requests import Response
+import CWMSpy  # type: ignore
+
+
+class ShefTransform(NamedTuple):
+    location: str
+    parameter_code: str
+    timeseries_id: str
+    units: Optional[str]
+    timezone: Optional[str]
+    dl_time: Optional[bool]
+
+
+class CdaValue(NamedTuple):
+    timestamp: int
+    value: float
+    quality: int
+
+
+TimeseriesPayload = TypedDict(
+    "TimeseriesPayload",
+    {"name": str, "office-id": str, "units": Optional[str], "values": list[CdaValue]},
+)
+
+CWMS_INTERVAL_PATTERN: str = r"([0-9]+)(\w+)"
+
+CWMS_INTERVAL_SECONDS: dict[str, int] = {
+    "Second": 1,
+    "Seconds": 1,
+    "Minute": 60,
+    "Minutes": 60,
+    "Hour": 3600,
+    "Hours": 3600,
+    "Day": 86400,
+    "Days": 86400,
+}
+
+
+class CdaLoader(base_loader.BaseLoader):
+    """
+    Loader used by cwms-data-api (CDA)
+    """
+
+    def __init__(
+        self,
+        logger: Optional[Logger],
+        output_object: Optional[Union[BufferedRandom, TextIO, str]] = None,
+        append: bool = False,
+    ) -> None:
+        """
+        Constructor
+        """
+        super().__init__(logger, output_object, append)
+        self._cda_url = "https://cwms-data-test.cwbi.us/cwms-data/"
+        self._cwms = CWMSpy.CWMS()
+        self._payloads: list[TimeseriesPayload] = []
+        self._time_series_error_count: int = 0
+        self._transforms: dict[str, ShefTransform] = {}
+        self._value_error_count: int = 0
+        self._write_tasks: list[Coroutine] = []
+
+    def set_options(self, options_str: str | None) -> None:
```

Comment:
Change to Union[str, None]

Reply:
Fixed in [8df6c69](https://github.com/HydrologicEngineeringCenter/SHEF_processing/pull/9/commits/8df6c697d07a19b6ab781208b1093360c09d13bb)